File: c++/023.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
//优先队列：存储每个链表的第一个结点及其值
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size()==0) return nullptr;
        if(lists.size()==1) return lists[0];
        //分冶法，类似于归并排序
        return merge(lists,0,lists.size()-1);
    }

    ListNode* merge(vector<ListNode*>& lists,int l, int h){
        if(l==h) return lists[l];
        if(l>h) return nullptr;
        int m = (l+h)>>1;
        return mergeTwo(merge(lists,l,m),merge(lists,m+1,h));
    }

    ListNode* mergeTwo(ListNode* A, ListNode* B){
        ListNode *h = new ListNode(0);
        ListNode *a=A, *b=B, *p=h;
        while(a&&b){
            if(a->val<b->val){
                p->next = a;
                p=a;
                a=a->next;
            }else{
                p->next = b;
                p = b;
                b = b->next;
            }
        }
        if(a) p->next = a;
        else p->next = b;
        return h->next;
    }
};

File: go/023.go
package main

/**
给你一个链表数组，每个链表都已经按升序排列。
请你将所有链表合并到一个升序链表中，返回合并后的链表。

思路：与21题类似
*/
func mergeKLists(lists []*ListNode) *ListNode {
	if len(lists) == 0 {
		return nil
	}
	if len(lists) == 1 {
		return lists[0]
	}
	c := lists[0]
	for i := 1; i < len(lists); i++ {
		c = _merge(c, lists[i])
	}
	return c
}

func _merge(p *ListNode, q *ListNode) *ListNode {
	if p == nil {
		return q
	}
	if q == nil {
		return p
	}

	if p.Val < q.Val {
		p.Next = _merge(p.Next, q)
		return p
	} else {
		q.Next = _merge(p, q.Next)
	}
	return q
}
